### ISSUE TITLE
Reactive Dict errors on clearing values that have never been retrieved

### ISSUE THREAD

**Issue by Pushplaybang** [meteor/meteor#5530]
Hi There

Found a weird one, I replicated in code, and have replicated simply in the chrome dev console so I don't believe a demo is necessary as its incredibly easy to replicate, but I can set one up if necessary.

With Reactive Dict included in your project doing the following in chrome dev tools

````js
var thing = new ReactiveDict();
thing.set('one', 'some value');
thing.clear();
````

produces this error : 

````js
reactive-dict.js:166 : Uncaught TypeError: Cannot read property '"some value"' of undefined(…)
(anonymous function)	@	reactive-dict.js:166
_.each._.forEach	@	underscore.js:142
_.extend.clear	@	reactive-dict.js:164
(anonymous function)	@	VM10168:2
InjectedScript._evaluateOn	@	VM9632:875
InjectedScript._evaluateAndWrap	@	VM9632:808
InjectedScript.evaluate	@	VM9632:664
````

If however you use the 'get'' prior to calling `.clear()` this does not occur, though calling .all() does not seem to stop the error for occurring when clearing afterwards.

**Comment by sprohaska** [comment]
I think the fix would be (sorry, I didn't have time to prepare a proper pull request):

```
diff --git a/packages/reactive-dict/reactive-dict.js b/packages/reactive-dict/reactive-dict.js
index d2b54b7..a807898 100644
--- a/packages/reactive-dict/reactive-dict.js
+++ b/packages/reactive-dict/reactive-dict.js
@@ -163,8 +163,10 @@ _.extend(ReactiveDict.prototype, {

     _.each(oldKeys, function(value, key) {
       changed(self.keyDeps[key]);
-      changed(self.keyValueDeps[key][value]);
-      changed(self.keyValueDeps[key]['undefined']);
+      if (self.keyValueDeps[key]) {
+        changed(self.keyValueDeps[key][value]);
+        changed(self.keyValueDeps[key]['undefined']);
+      }
     });

   },
```

**Comment by Pushplaybang** [comment]
Until this gets resolved, and for anyone else running into this issue, I'm using this as a hack in my app : 

````js
App.Ui = {
  //..
  reactiveDictClearHack: function(obj) {
    _.each(obj.formValues.all(), function(value, key) {
      obj.formValues.get(key);
    });
  }
}
````
and then call

```js
App.Ui.Util.reactiveDictClearHack(App.Ui.topics);
```

this is of course specific to my setup, where each obj on App.Ui contains a formValues reactive dictionary.

Looking forward to not having to include crap like this in my app.

**Comment by stubailo** [comment]
Filed a PR, will add a test then merge.

**Comment by Pushplaybang** [comment]
+1 @stubailo nice one!

**Comment by Crenshinibon** [comment]
What are the plans to make a new release including this fix? I don't want to create a workaround if the fix get's rolled out in a week. But if it takes longer I have to come up with something.

Is it actually possible (and how) to add core packages from the devel branch?

**Comment by Sojourneer** [comment]
Not fixed in Meteor 1.2.1. Lame.

**Comment by abernix** [comment]
@Sojourneer Meteor 1.2.1 (which is a year old next week) is the last version that this was broken in.  This is fixed in Meteor 1.3 (and higher) and Meteor 1.4.1.2 is out now.  Please consider upgrading!

**Comment by Sojourneer** [comment]
I'm scared spitless to upgrade. Once I figure out how to do automated testing for my app (first production app), I'll consider doing so.
